DOC: correct, update, and extend `lobpcg` docstring info and examples

lobpcg opened the pull request:
ENH: update and extend lobpcg.py docstring info and examples

#### Reference issue
N/A

#### What does this implement/fix?
update and extend lobpcg.py docstring info and examples

lobpcg commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,57 +155,55 @@ def lobpcg(
+        of the column-space of Y. Y must be full rank if present.
+    tol : scalar, optional. The default is ``tol=n*sqrt(eps)``.
+        Solver tolerance for the stopping criterion.
+    maxiter : int, optional. The default is ``maxiter=20``.
+        Maximum number of iterations.
```
Already done!

lobpcg replied:
Or maybe I do not understand your request...

melissawm replied:
Maybe it's not necessary but I would add something like 
```suggestion
    maxiter : int, optional. The default is ``maxiter=20``.
        Maximum number of iterations. The actual number of iterations will be ``maxiter+3`` (see Notes section).
```

lobpcg replied:
The extra output in history does not come from iterations, but rather from pre- and post-processing.

melissawm replied:
Oh, got it. Thanks for clarifying!

melissawm commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +271,111 @@ def lobpcg(
+
+    >>> vals = np.arange(1, n + 1).astype(np.int16)
+
+    The first mandatory input parameter, in this test is
```
```suggestion
    The first mandatory input parameter in this test is
```

melissawm reviewed (commented):
Left a couple minor comments, thanks @lobpcg !

melissawm reviewed (approved):
This looks good to me now! I'll leave this for other maintainers to see, if we don't get other feedback I'll merge in a few days.

melissawm commented:
Single quotes will create links whenever possible to existing objects (e. g. `np.array` will create a link to the array doc). Double quotes are used for raw formatting. If you use single quotes and a link is not found, the result is left in italics.

lobpcg commented:
> Single quotes will create links whenever possible to existing objects (e. g. `np.array` will create a link to the array doc). Double quotes are used for raw formatting. If you use single quotes and a link is not found, the result is left in italics. 

Thanks! I'll make the necessary changes and massage the new example to pass the test soon

lobpcg commented:
@melissawm I have added another short example at the end and fixed, I hope, all formatting related to quotes, after your review. 

It's ready to merge.

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
-        of the column-space of Y. Y must be full rank.
-    tol : scalar, optional.
-        Solver tolerance (stopping criterion).
+        Initial approximation to the ``k`` eigenvectors (non-sparse).
+        If ``A`` has ``shape=(n,n)`` then ``X`` must have ``shape=(n,k)``.
```
Parameters also need a single quote.
```suggestion
        Initial approximation to the ``k`` eigenvectors (non-sparse).
        If `A` has ``shape=(n,n)`` then `X` must have ``shape=(n,k)``.
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+        By default ``Y = None``.
+        An ``n-by-sizeY`` ndarray of constraints with ``sizeY < n``.
+        The iterations will be performed in the ``B``-orthogonal complement
+        of the column-space of ``Y``. ``Y`` must be full rank if present.
```
```suggestion
        of the column-space of `Y`. `Y` must be full rank if present.
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+        Solver tolerance for the stopping criterion.
+    maxiter : int, optional
+        The default is ``maxiter=20``. Maximum number of iterations.
+    largest : bool, optional. The default is ``largest=True``.
```
This is not conform to numpydoc, see https://numpydoc.readthedocs.io/en/latest/format.html#parameters
```suggestion
    largest : bool, default: True
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
-        Whether to return history of residual norms.  Default is False.
+        By default ``verbosityLevel=0`` no output.
+        Controls the solver standard/screen output.
+    retLambdaHistory : bool, optional
+        The default is ``False``.
```
Then we should be consistent
```suggestion
    retLambdaHistory : bool, default: False
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+    retLambdaHistory : bool, optional
+        The default is ``False``.
+        Whether to return iterative eigenvalue history.
+    retResidualNormsHistory : bool, optional
+        The default is ``False``.
```
```suggestion
    retResidualNormsHistory : bool, default: False
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+    v : ndarray of the same shape as ``X.shape``.
+        An array of ``k`` approximate eigenvectors.
+    lambdaHistory : ndarray, optional.
+        The eigenvalue history, if ``retLambdaHistory`` is ``True``.
+    ResidualNormsHistory : ndarray, optional.
+        The history of residual norms, if ``retResidualNormsHistory``
```
```suggestion
        The eigenvalue history, if `retLambdaHistory` is ``True``.
    ResidualNormsHistory : ndarray, optional.
        The history of residual norms, if `retResidualNormsHistory`
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
     than the last one iterated, as a cure for possible divergence.
 
+    If ``X.dtype == np.float32`` and user-provided operations/multiplications
+    by ``A``, ``B``, and ``M`` all presere the ``np.float32`` data type,
```
```suggestion
    by `A`, `B`, and `M` all preserve the ``np.float32`` data type,
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+    (limited by `maxit`) iterations plus 3: initial, final, and postprocessing.
 
-    If both ``retLambdaHistory`` and ``retResidualNormsHistory`` are True,
+    If both ``retLambdaHistory`` and ``retResidualNormsHistory`` are ``True``,
```
```suggestion
    If both `retLambdaHistory` and `retResidualNormsHistory` are ``True``,
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -236,20 +244,21 @@ def lobpcg(
-    The convergence speed depends basically on two factors:
+    The convergence speed depends basically on three factors:
+
+    1. Quality of the initial approximations ``X`` to the seeking eigenvectors.
```
```suggestion
    1. Quality of the initial approximations `X` to the seeking eigenvectors.
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -236,20 +244,21 @@ def lobpcg(
-    2. Quality of the initial approximations `X` to the seeking eigenvectors.
-       Randomly distributed around the origin vectors work well if no better
-       choice is known.
+       for ``A``, which is easy to code since ``A`` is tridiagonal.
```
```suggestion
       for `A`, which is easy to code since `A` is tridiagonal.
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
+    >>> vals = np.arange(1, n + 1).astype(np.int16)
+
+    The first mandatory input parameter in this test is
+    the sparse diagonal matrix ``A``
```
```suggestion
    the sparse diagonal matrix `A`
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
 
-    Initial guess for eigenvectors, should have linearly independent
-    columns. The second mandatory input parameter, a 2D array with the
+    The second mandatory input parameter ``X`` is a 2D array with the
```
```suggestion
    The second mandatory input parameter `X` is a 2D array with the
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
-    columns. The second mandatory input parameter, a 2D array with the
+    The second mandatory input parameter ``X`` is a 2D array with the
     row dimension determining the number of requested eigenvalues.
+    ``X`` is an initial guess for targeted eigenvectors.
+    ``X`` must have linearly independent columns.
```
```suggestion
    `X` is an initial guess for targeted eigenvectors.
    `X` must have linearly independent columns.
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
+    >>> eigenvalues.dtype
+    dtype('float32')
+
+    LOBPCG needs only access the matrix product with ``A`` rather
+    then the matrix itself. Since the matrix ``A`` is diagonal in
```
```suggestion
    LOBPCG needs only access the matrix product with `A` rather
    then the matrix itself. Since the matrix `A` is diagonal in
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
     >>> Y = np.eye(n, 3)
 
-    Preconditioner in the inverse of A in this example:
+    The preconditioner acts as the inverse of ``A`` in this example, but
+    in the reduced precision ``np.float32`` even though the initial ``X``
```
```suggestion
    The preconditioner acts as the inverse of `A` in this example, but
    in the reduced precision ``np.float32`` even though the initial `X`
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
+    >>> M = lambda X: inv_vals[:, np.newaxis] * X
 
-    The preconditiner must be defined by a function:
+    Let us now solve the eigenvalue problem for the matrix ``A`` first
```
```suggestion
    Let us now solve the eigenvalue problem for the matrix `A` first
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
 
-    The argument x of the preconditioner function is a matrix inside `lobpcg`,
-    thus the use of matrix-matrix product ``@``.
+    With preconditioning we need only 20 iterations from the same ``X``
```
```suggestion
    With preconditioning we need only 20 iterations from the same `X`
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
 
-    >>> M = LinearOperator(matvec=precond, matmat=precond,
-    ...                    shape=(n, n), dtype=np.float64)
+    Note that the vectors passed in ``Y`` are the eigenvectors of the 3
```
```suggestion
    Note that the vectors passed in `Y` are the eigenvectors of the 3
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -268,71 +277,125 @@ def lobpcg(
+    smallest eigenvalues. The results returned above are orthogonal to those.
 
-    Let us now solve the eigenvalue problem for the matrix A:
+    Finally, the primary matrix ``A`` may be indefinite, e.g., after shifting
```
```suggestion
    Finally, the primary matrix `A` may be indefinite, e.g., after shifting
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
+    M : {sparse matrix, ndarray, LinearOperator, callable object}
+        Optional. By default ``M = None``, which is equivalent to identity.
+        Preconditioner aiming to accelerate convergence.
+    Y : ndarray, float32 or float64, optional
+        By default ``Y = None``.
```
```suggestion
    Y : ndarray, float32 or float64, default: None
```

tupui commented on scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:
```diff
@@ -155,71 +155,79 @@ def lobpcg(
-        Maximum number of iterations.  The default is ``maxiter=20``.
-    largest : bool, optional.
+        Solver tolerance for the stopping criterion.
+    maxiter : int, optional
+        The default is ``maxiter=20``. Maximum number of iterations.
```
```suggestion
    maxiter : int, default: 20
        Maximum number of iterations.
```

tupui reviewed (changes requested):
Thanks for the addition @lobpcg 

@melissawm Single quotes are also needed when talking about a parameter of the function/class/method. Like ``` `A` ``` but when using it we need double quotes like ``` ``A=...`` ```.

If you accept the suggestions. Please use the `Files changed` tabs and click on `Add suggestion to batch`. This way it will batch all changes and create a single commit, hence send a single notification. And please in the commit message add `[skip azp] [skip actions] [skip cirrus]`

lobpcg commented:
@tupui sorry, forgot the commit message add [skip azp] [skip actions] [skip cirrus]...

tupui reviewed (approved):
Thank you for the update @lobpcg (no worries for the CI, next time 😉).

I am merging since there is already another approval from a maintainer.
